### PR TITLE
Parenthesize breaking named expressions in match guards

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This is an attempt to solve https://github.com/astral-sh/ruff/issues/9394 by avoiding breaks in named expressions when invalid.

### CODE REVIEW

File: crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap [hunk 1]
```diff
@@ -677,7 +682,9 @@ match newlines:
 
 
 match long_lines:
-    case "this is a long line for if condition" if aaaaaaaaahhhhhhhh == 1 and bbbbbbaaaaaaaaaaa == 2:  # comment
+    case "this is a long line for if condition" if (
+        aaaaaaaaahhhhhhhh == 1 and bbbbbbaaaaaaaaaaa == 2
+    ):  # comment
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap:687]: This is now parenthesized, whereas it wasn't before...

File: crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap [hunk 2]
```diff
@@ -696,6 +703,11 @@ match long_lines:
     ):  # another comment
         pass
 
+    case {"long_long_long_key": str(long_long_long_key)} if (
+        value := "long long long long long long long long long long long value"
+    ):
+        pass
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@statement__match.py.snap:709]: Black formats as:

```python
    case {
        "long_long_long_key": str(long_long_long_key)
    } if value := "long long long long long long long long long long long value":
        pass
```

**Review by MichaReiser — COMMENTED** [review]
I feel the same as you. Maybe parenthesze is mainly intended to use for top level expressions, not for nested sub expressions.

I haven't looked at the issue (on my phone) but this is probably something that needs solving in the case formatting or we should use in_parentheses_only_softline_break in the if expression formatting

**Comment by charliermarsh** [comment]
Let me try `in_parentheses_only_softline_break`, that's interesting.

**Comment by charliermarsh** [comment]
Ok this now looks right...

**Comment by zanieb** [comment]
Nice!

File: crates/ruff_python_formatter/src/expression/expr_named_expr.rs [hunk 3]
```diff
@@ -24,7 +26,10 @@ impl FormatNodeRule<ExprNamedExpr> for FormatExprNamedExpr {
         write!(
             f,
             [
-                group(&format_args!(target.format(), soft_line_break_or_space())),
+                group(&format_args!(
+                    target.format(),
+                    in_parentheses_only_soft_line_break_or_space()
+                )),
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/expression/expr_named_expr.rs:32]: Nit
```suggestion
                group(&format_args![
                    target.format(),
                    in_parentheses_only_soft_line_break_or_space()
                ]),
```

**Review by MichaReiser — APPROVED** [review]

**Comment by MichaReiser** [comment]
Thank you

**Pull request merged**